web-common/src/features/scheduled-reports/utils.ts

```diff
@@ -17,3 +67,93 @@ export function getDashboardNameFromReport(
     null
   );
 }
+
+export function extractNotification(
+  notifiers: V1Notifier[] | undefined,
+  userEmail: string | undefined,
+  isEdit: boolean,
+): AlertNotificationValues {
+  const slackNotifier = notifiers?.find((n) => n.connector === "slack");
+  const slackChannels = mapAndAddEmptyEntry(
+    slackNotifier?.properties?.channels as string[] | undefined,
+    "channel",
+  );
+  const slackUsers = mapAndAddEmptyEntry(
+    slackNotifier?.properties?.users as string[] | undefined,
+    "email",
+  );
+
+  const emailNotifier = notifiers?.find((n) => n.connector === "email");
+  const emailRecipients = mapAndAddEmptyEntry(
+    emailNotifier?.properties?.recipients as string[] | undefined,
+    "email",
+  );
+
+  if (userEmail && !isEdit) {
+    slackUsers.push({
+      email: userEmail,
+    });
+    emailRecipients.push({
+      email: userEmail,
+    });
+  }
+
+  return {
+    enableSlackNotification: isEdit ? !!slackNotifier : true,
+    slackChannels,
+    slackUsers,
+
+    enableEmailNotification: isEdit ? !!emailNotifier : true,
+    emailRecipients,
+  };
+}
+
+export function extractNotificationV2(
+  notifiers: V1Notifier[] | undefined,
+  userEmail: string | undefined,
+  isEdit: boolean,
+) {
+  const slackNotifier = notifiers?.find((n) => n.connector === "slack");
+  const slackChannels = mapAndAddEmptyEntryV2(
+    slackNotifier?.properties?.channels as string[],
+  );
+  const slackUsers = mapAndAddEmptyEntryV2(
+    slackNotifier?.properties?.users as string[],
+  );
+
+  const emailNotifier = notifiers?.find((n) => n.connector === "email");
+  const emailRecipients = mapAndAddEmptyEntryV2(
+    emailNotifier?.properties?.recipients as string[],
+  );
+
+  if (userEmail && !isEdit) {
+    slackUsers.unshift(userEmail);
+    emailRecipients.unshift(userEmail);
+  }
+
+  return {
+    enableSlackNotification: isEdit ? !!slackNotifier : true,
+    slackChannels,
+    slackUsers,
+
+    enableEmailNotification: isEdit ? !!emailNotifier : true,
+    emailRecipients,
+  };
+}
+
+function mapAndAddEmptyEntry<K extends string>(
+  entries: string[] | undefined,
+  key: K,
+) {
+  const mappedEntries = entries?.map((e) => ({ [key]: e })) ?? [];
+  mappedEntries.push({ [key]: "" });
+  return mappedEntries as {
+    [KEY in K]: string;
+  }[];
+}
+
+function mapAndAddEmptyEntryV2(entries: string[] | undefined) {
+  const finalEntries = entries ? [...entries] : [];
+  finalEntries.push("");
+  return finalEntries;
+}
```

Comment:
Do we need to add an empty entry? It'd be simpler if we didn't.

Reply:
The empty entry is what is edited. It is mainly to support submit without needing to convert the currently being edited entry to a pill explicitly.